gc/base/standard/ConcurrentGC.cpp

```diff
@@ -1581,8 +1581,11 @@ MM_ConcurrentGC::signalThreadsToActivateWriteBarrier(MM_EnvironmentBase *env)
 	while (CONCURRENT_INIT_COMPLETE == _stats.getExecutionMode()) {
 		/* We may or may not have exclusive access but another thread may have beat us to it and
 		 * prepared the threads or even collected.
+		 *
+		 * SATB marks all newly allocated objectes during active concurrent cycle.
+		 * Since it's done on TLH granularity we have to flush the current ones and start creating new ones, once the cycle starts.
 		 */
-		if (env->acquireExclusiveVMAccessForGC(this, true, false)) {
+		if (env->acquireExclusiveVMAccessForGC(this, true, _extensions->usingSATBBarrier())) {
```

Comment:
Possibly excessive, but you could have a virtual method MM_ConcurrentGC::acquireExclusiveVMAccessForGC()
that does 

env->acquireExclusiveVMAccessForGC(this, true, true/false) 
for SATB/IU variants.

Now, we are talking OO :)

Reply:
However, I would not extend that logic on similar spots in Sweep and HeapWalker, since I don't want to introduce SATB/UI subclasses just for that one call, but MM_ConcurrentGC already has those subclasses, so it's easy to utilize it.